### ISSUE TITLE
Template caching question

### ISSUE THREAD

**Issue by wjthomas9** [devoptix/gulp-nunjucks-api#2]
Hi again.

When using gulp-data to read data from a json file, I'm just curious if you know of a way to force nunjucks variables to read the updated data without having to stop and start gulp. Maybe this is a question for gulp-data, since it's the module that uses the require function.

**Comment by wjthomas9** [comment]
I made the following change and it seems to work.

```
function getDataForFile(file){
    // return require('./build/data/' + path.basename(file.path, '.html') + '.json');
    return JSON.parse(fs.readFileSync('./build/data/' + path.basename(file.path, '.html') + '.json'));
}
```

**Comment by waynebloss** [comment]
Thanks for bringing this to my attention. That will work for all of the json and js files that you supply.

I will have to change it for "locals" files, which are json and js files that are pulled into the context for one specific file. I will make it use [require-new](https://www.npmjs.com/package/require-new) instead of require [here](https://github.com/devoptix/gulp-nunjucks-api/blob/master/index.js#L152).

**Comment by wjthomas9** [comment]
I appreciate the help.

Just for some context, I'm moving from Gulp Swig to trying to use Nunjucks for my frontend prototyping. I don't use it with Express.js, but it a handy cache: false option that I can use in the gulp task, so that the data will be refreshed when you watch files.

http://paularmstrong.github.io/swig/docs/api/#SwigOpts

Unfortunately, Paul Armstrong announced that Swig is not currently being maintained, and suggested Nunjucks.

https://github.com/paularmstrong/swig/issues/628

**Comment by wjthomas9** [comment]
Oh, I see. I was deliberately removing the .html portion of the extension for my data files.

**Comment by waynebloss** [comment]
It doesn't matter what you name your data files if you're loading them yourself with gulp-data.

If you want to use the "locals" system, which is a little different than gulp-data, then you have to follow the default pattern of `<filename>.+(js|json)`. The locals configuration option can also be set to a different pattern such as `<filename_noext>.+(js|json)`. I don't think you're using the locals system though.

**Comment by waynebloss** [comment]
However, I would not use the locals system until I have made the change to using require-new.

**Comment by wjthomas9** [comment]
Oh ok.  I see. I’m not familiar with the locals system and how it works, so I’ll have to do some learning.

**Comment by waynebloss** [comment]
No problem. I just fixed locals so that it uses require-new, so that it will reload each time the template is rendered. So if you want to use it, it's fixed for your use-case now in v0.6.3.

To use it, you just set `locals: true` in the gulp task and then if you have a template file named `index.html`, when it renders that file, it will look for any files named `index.html.js` and `index.html.json` and merge the data and functions from those files into the rendering context so you can use them in your template. This is useful if you have some piece of data or functionality that is specific or "local" to a single template. Also, note that It only works for top-level templates, not templates that you would extend or include.

**Comment by wjthomas9** [comment]
Great! Can the .js and .json files be in subfolders?

**Comment by waynebloss** [comment]
Yes I think so but I haven't tested it. To try it out, instead of setting `locals: true` you would set `locals: './subfolder/<filename>.+(js|json)'`.

If it doesn't work, we can make a new issue for that.

**Comment by waynebloss** [comment]
Actually I'm not sure that starting the path with a dot is necessary, it might have to be `locals: 'subfolder/<filename>.+(js|json)'` but you can play around with it and see the other rules for paths at  [isaacs/node-glob*#*glob-primer](https://github.com/isaacs/node-glob#glob-primer).

Also, I just released v0.6.4 which lets you refer to the root *src* directory as `/` in the locals glob path.

**Comment by wjthomas9** [comment]
Great. I really appreciate all the input. Thanks for taking the time.

On Jul 2, 2015, at 7:01 PM, Wayne Bloss <notifications@github.com> wrote:

Actually I'm not sure that starting the path with a dot is necessary, it
might have to be locals: 'subfolder/<filename>.+(js|json)' but you can play
around with it and see the other rules for paths at isaacs/node-glob*#*
glob-primer <https://github.com/isaacs/node-glob#glob-primer>.

Also, I just released v0.6.4 which lets you refer to the root *src*
directory as / in the locals glob path.

—
Reply to this email directly or view it on GitHub
<https://github.com/devoptix/gulp-nunjucks-api/issues/2#issuecomment-118197382>
.

**Comment by waynebloss** [comment]
Note to readers: I actually reverted the "root src directory" behavior for locals because having `/` resolve to a single path doesn't work when people provide multiple src directories in an array as seen in #3.

The locals system is my own invention and I'm not sure how useful it will be for anyone, but I guess we'll find out.
